functions/cache/src/async_cache.cpp

```diff
@@ -0,0 +1,422 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "kvs_async_client.hpp"
+#include "yaml-cpp/yaml.h"
+
+ZmqUtil zmq_util;
+ZmqUtilInterface* kZmqUtil = &zmq_util;
+
+unsigned kCacheReportThreshold = 5;
+
+struct PendingClientMetadata {
+  PendingClientMetadata() = default;
+  PendingClientMetadata(set<Key> read_set, set<Key> to_cache_set) :
+      read_set_(std::move(read_set)),
+      to_cache_set_(std::move(to_cache_set)) {}
+  set<Key> read_set_;
+  set<Key> to_cache_set_;
+};
+
+string get_serialized_value_from_cache(
+    const Key& key, LatticeType type,
+    const map<Key, LWWPairLattice<string>>& local_lww_cache,
+    const map<Key, SetLattice<string>>& local_set_cache, logger log) {
+  if (type == LatticeType::LWW) {
+    if (local_lww_cache.find(key) != local_lww_cache.end()) {
+      return serialize(local_lww_cache.at(key));
+    } else {
+      log->error("Key {} not found in LWW cache.", key);
+      return "";
+    }
+  } else if (type == LatticeType::SET) {
+    if (local_set_cache.find(key) != local_set_cache.end()) {
+      return serialize(local_set_cache.at(key));
+    } else {
+      log->error("Key {} not found in SET cache.", key);
+      return "";
+    }
+  } else {
+    log->error("Invalid lattice type.");
+    return "";
+  }
+}
+
+void update_cache(const Key& key, LatticeType type, const string& payload,
+                  map<Key, LWWPairLattice<string>>& local_lww_cache,
+                  map<Key, SetLattice<string>>& local_set_cache, logger log) {
+  if (type == LatticeType::LWW) {
+    local_lww_cache[key].merge(deserialize_lww(payload));
+  } else if (type == LatticeType::SET) {
+    local_set_cache[key].merge(deserialize_set(payload));
+  } else {
+    log->error("Invalid lattice type.");
+  }
+}
+
+void send_get_response(const set<Key>& read_set, const Address& response_addr,
+                       const map<Key, LatticeType>& key_type_map,
+                       const map<Key, LWWPairLattice<string>>& local_lww_cache,
+                       const map<Key, SetLattice<string>>& local_set_cache,
+                       SocketCache& pushers, logger log) {
+  KeyResponse response;
+  response.set_type(RequestType::GET);
+  for (const Key& key : read_set) {
+    KeyTuple* tp = response.add_tuples();
+    tp->set_key(key);
+    tp->set_lattice_type(key_type_map.at(key));
+    // check if key exists
+    if (local_lww_cache.find(key) == local_lww_cache.end() &&
```

Comment:
in the current code this won't really work because `key_type_map` is being set even if the key doesn't exist in both lww and set caches...The reason is that we require client to specify a lattice type even during the GET request, and we are updating the `key_type_map` at that point. I modified the logic such that the client doesn't specify a lattice type during GET, and I also changed some other code such that now our `key_type_map` only contains keys that are actually cached. See the new code and let me know how you think.

Reply:
Yes, I think the new implementation is much cleaner.